src/package-hoister.js

```diff
@@ -196,7 +200,12 @@ export default class PackageHoister {
 
       for (const depPattern of ref.dependencies) {
         const depinfo = this._lookupDependency(info, depPattern);
-        if (depinfo && !depinfo.isRequired && !depinfo.isIncompatible) {
+        if (
+          depinfo &&
+          !depinfo.isRequired &&
+          !depinfo.isIncompatible &&
+          (!depinfo.pkg._reference || !depinfo.pkg._reference.optional || !this.ignoreOptional)
```

Comment:
We can defer this to a follow-up since it's not the main goal of the PR anyways.